File: 380-insert-delete-getrandom-o1/solution.ts
class RandomizedSet {
  private map = new Map<number, number>();
  private arr: number[] = [];
  private del: number = 0;

  insert(val: number): boolean {
    if (!this.map.has(val)) {
      this.map.set(val, this.arr.length);
      this.arr.push(val);
      return true;
    }
    return false;
  }

  remove(val: number): boolean {
    const idx = this.map.get(val);
    if (idx !== undefined) {
      this.map.set(this.arr[this.del], idx);
      [this.arr[idx], this.arr[this.del]] = [this.arr[this.del], this.arr[idx]];
      this.del++;
      this.map.delete(val);
      return true;
    }
    return false;
  }

  getRandom(): number {
    const rand =
      this.del + Math.floor(Math.random() * (this.arr.length - this.del));
    return this.arr[rand];
  }
}

File: 380-insert-delete-getrandom-o1/solution.cpp
class RandomizedSet {
public:
    mt19937 engine{random_device{}()};
    unordered_map<int, int> mp;
    vector<int> v;

    RandomizedSet() {}
    
    bool insert(int val) {
        if(mp.contains(val)) return false;
        mp[val] = v.size();
        v.push_back(val);
        return true;
    }
    
    bool remove(int val) {
        if(!mp.contains(val)) return false;
        int idx = mp[val];
        if(idx != v.size() - 1) {
            v[idx] = v.back();
            mp[v.back()] = idx;
        }
        v.pop_back();
        mp.erase(val);
        return true;
    }
    
    int getRandom() {
        int n = v.size();
        uniform_int_distribution<> dist(0, n - 1);
        int idx = dist(engine);
        return v[idx];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */
